news/_posts/2019-09-11-prestashop-in-2019-and-beyond-part-2-pain-points.md

```diff
@@ -0,0 +1,471 @@
+
+#### Technical debt
+
+Technical debt is a problem for all mature projects, and PrestaShop is not the exception there. As stated before, the best way to avoid software rot is to apply frequent refactoring. But if refactoring is limited by breaking changes, then it cannot be applied properly and technical debt starts piling up.
+
+Even though continuous efforts are being made to avoid this, having to circle around breaking changes ultimately increases complexity. And indeed, the current architecture has become quite complex, with a legacy part that contains a great number of crisscrossing dependencies, and a new-generation part that still requires following many not-so-obvious rules and many workarounds. In addition, more layers means more code to load in memory, which in some cases can also result in a performance penalty.
+
+
+## No clearly-defined API
+
+As we said before, interoperability is a big subject in platform development. Since refactoring is, in the long term, unavoidable, breaking changes are bound to occur sooner or later.
+
+Remember back in 2015, when [PrestaShop started following SemVer][semver-article]? The Semantic Versioning (SemVer) [specification](https://semver.org/) defines that only major versions can introduce backwards incompatible API changes, while minor and patch versions must ensure backwards compatibility. 
+
+This is good for interoperability: it means that as long as integrations respect the software's API, they can be sure they will keep working with the main software, even if that software is upgraded to the next minor or patch version.
+
+In the case of PrestaShop, merchants want to be able to upgrade their shop seamlessly and benefit from the new features and bug fixes, without having to change their theme, modules and customizations, or fear something will break down. As stated before, SemVer is supposed to ensure this, as long as third parties respect PrestaShop's API. This is true—in theory.
+
+**But... what _exactly_ constitutes PrestaShop's API?**
+
+Right now there's no article documenting it out anywhere. If there was, we would be able to say: "This is PrestaShop's API. It is something that we are committing on. Respect it, and enjoy forward compatibility!".
+
+So, if it's not clearly defined anywhere... then what is this unwritten contract made up from? Let's find out.
+
+First, we must first identify **who** are our technical stakeholders. Then, we have to establish **what** are the technical resources they need to interact with, and finally define **how** those interactions will take place. Sounds familiar? We have already depicted that in the [previous article][part-1] (see [figure][comprehensive-overview]).
```

Comment:
Technical stakeholder seems a technical term to me 🤔. We can use it but I'd suggest we provide a short definition to help people who have never heard about it in software development context before.

Reply:
I think it's self explanatory: who needs what and how. What do you suggest?

Reply:
Maybe a link to https://en.wikipedia.org/wiki/Stakeholder_management ?

It's just that, before working at PrestaShop, I never encountered this word in a software development context, so the 1st time I read it, I was puzzled. Then it started to make sense when I thought about the word and the context. But I had a "😮" moment that some readers might experience too

Reply:
Interesting...

WDYT?

```suggestion
First, we must first identify **who** are our technical [stakeholders](https://en.wikipedia.org/wiki/Stakeholder_management). Then, we have to establish **what** are the technical resources they need to interact with, and finally define **how** those interactions will take place. Sounds familiar? We have already depicted that in the [previous article][part-1] (see [figure][comprehensive-overview]).
```

Reply:
👍